inc/sitemaps/class-post-type-sitemap-provider.php

```diff
@@ -471,31 +459,37 @@ protected function get_first_links( $post_type ) {
 	 */
 	protected function get_post_type_archive_link( $post_type ) {
 
-		if ( WPSEO_Options::get( 'noindex-ptarchive-' . $post_type, false ) ) {
-			return false;
-		}
+		$pt_archive_page_id  = -1;
+		$show_posts_on_front = false;
 
-		// Post archive should be excluded if it isn't front page or posts page.
-		if ( $post_type === 'post' && get_option( 'show_on_front' ) !== 'posts' && ! $this->get_page_for_posts_id() ) {
-			return false;
+		if ( $post_type === 'post' ) {
+			$pt_archive_page_id  = get_option( 'page_for_posts' );
+			$show_posts_on_front = ( get_option( 'show_on_front' ) === 'posts' );
+
+			// Post archive should be excluded if it isn't front page or posts page.
+			if ( ! $show_posts_on_front && $pt_archive_page_id <= 0 ) {
+				return false;
+			}
 		}
 
 		/**
 		 * Filter the page which is dedicated to this post type archive.
 		 *
-		 * @param string $post_id   The post_id of the page.
-		 * @param string $post_type The post type this archive is for.
+		 * @param string $pt_archive_page_id The post_id of the page.
+		 * @param string $post_type          The post type this archive is for.
 		 */
-		$post_id = (int) apply_filters(
-			'wpseo_sitemap_page_for_post_type_archive',
-			( 'post' === $post_type ) ? $this->get_page_for_posts_id() : ( -1 ),
-			$post_type
-		);
+		$pt_archive_page_id = (int) apply_filters( 'wpseo_sitemap_page_for_post_type_archive', $pt_archive_page_id, $post_type );
 
-		if ( $post_id > 0 && WPSEO_Meta::get_value( 'meta-robots-noindex', $post_id ) === '1' ) {
+		if ( WPSEO_Options::get( 'noindex-ptarchive-' . $post_type, false ) ||
+			( $pt_archive_page_id > 0 && WPSEO_Meta::get_value( 'meta-robots-noindex', $pt_archive_page_id ) === '1' )
+		) {
 			return false;
 		}
 
+		if ( $post_type === 'post' && $show_posts_on_front ) {
```

Comment:
Wouldn't it be more sensible to move this a bit above inside the other `if` and then extract that to a separate method?

Reply:
I've tried couple variants and I've created new method _is_post_type_archive_indexable_. On this way, we can reduce complexity of the code and make it more readable.

I assume that if the homepage shows latest posts then the homepage can't be excluded (I think that there isn't "noindex" setting in the UI).